### ISSUE TITLE
Sequence Points

### ISSUE THREAD

**Issue by KirillOsenkov** [icsharpcode/ILSpy#1041]
If I recall correctly the old version of ICSharpCode.Decompiler.dll provided a way to get the sequence points for decompiled methods when writing the tree out to text.

I do see the CreateSequencePoints() method on CSharpDecompiler, however it doesn't seem very useful as the tree it is running on doesn't have line/column information (StartLocation/EndLocation are empty). My understanding is that AstNodes are not full fidelity like Roslyn (they don't preserve exact whitespace/trivia) and so it is up to the text writer to determine trivia/whitespace. This is why the sequence points only make sense when we actually output the tree to text.

I've added a simple unit-test to test CreateSequencePoints:
https://github.com/icsharpcode/ILSpy/pull/1040
It's not necessarily to be checked in, but I thought it'd be useful to have a test like this.

This is not urgent and not blocking, but I'd be interested to see if anyone has any thoughts on how to properly implement sequence points. What I need is to decompile a type to text with all members, and to be able to get sequence points with valid locations in that text for each member. I've attempted to write a visitor (like a Roslyn tree rewriter) that would assign locations to nodes, but quickly realized that even if I hack this together (concatenate lengths of all terminal nodes in the tree) it won't necessarily match the lines/columns present in the text output.

It would feel weird to couple the text output machinery with sequence points, but I don't see any other easy way.

Maybe write the tree to text, then parse it back, this time filling out locations? Then we'd need a C# parser. Or, parse into a Roslyn syntax tree, and have a visitor to create an AstNode-based tree from a Roslyn syntax tree, with all locations preserved.

Again, this is not urgent and not required for MonoDevelop. I'm just playing with ILSpy. BTW ILSpy is awesome and thank you all so much for your efforts, it is much appreciated.

**Comment by siegfriedpammer** [comment]
>Maybe write the tree to text, then parse it back, this time filling out locations? Then we'd need a C# parser. Or, parse into a Roslyn syntax tree, and have a visitor to create an AstNode-based tree from a Roslyn syntax tree, with all locations preserved.

We already have functionality to do this, without parsing the text back into an AST. See this snippet:
https://github.com/icsharpcode/ILSpy/blob/4cccf9fac59e20434896646025848774da02d110/ILSpy/Languages/CSharpILMixedLanguage.cs#L46-L52

#1040 is failing... is this on purpose?

**Comment by KirillOsenkov** [comment]
Ah, thanks, that's exactly the missing piece I've been looking for. The test fails in Debug because it asserts that the tree has locations, and it doesn't.

I've added the code above to the test and it now works great! I will update my PR shortly.

One thing I've noticed is that your snippet above is a bit redundant: WrapInWriterThatSetsLocationsInAST does this:
https://github.com/icsharpcode/ILSpy/blob/0cd8e21ff7cdbfa30875a669480c3c5474ac1d5d/ICSharpCode.Decompiler/CSharp/OutputVisitor/ITokenWriter.cs#L75

Basically it adds InsertSpecials, InsertRequiredSpaces and InsertMissingTokens.

The CSharpOutputVisitor ctor used does this:
https://github.com/icsharpcode/ILSpy/blob/0cd8e21ff7cdbfa30875a669480c3c5474ac1d5d/ICSharpCode.Decompiler/CSharp/OutputVisitor/CSharpOutputVisitor.cs#L60

It also adds InsertSpecials and InsertRequiredSpaces. One of those doesn't need to add those two which are duplicated.

**Comment by KirillOsenkov** [comment]
Perfect, I've updated #1040 with the information you gave me and it seems to work great! That would be a good sample of how to get sequence points.

I have to admit that I've read the code and I still don't understand at what point the syntax tree gets mutated by the CSharpOutputVisitor to contain the TextLocations.

Also see my comment above regarding duplicated InsertSpecialsDecorator and InsertRequiredSpacesDecorator - you decide if it's a bug that you want to fix, I can file it if needed.

**Comment by siegfriedpammer** [comment]
> Also see my comment above regarding duplicated InsertSpecialsDecorator and InsertRequiredSpacesDecorator - you decide if it's a bug that you want to fix, I can file it if needed.

Yes, I would say this is a thing we could improve. You can file an issue for it. Thanks!
